File: android/app/src/main/java/com/ciro/app/data/model/Incident.kt
package com.ciro.app.data.model

/**
 * Kotlin data classes mirroring the Firestore `incidents` collection schema.
 * TEAM CONTRACT — matches the Python Pydantic models exactly.
 */
data class Incident(
    val incident_id: String = "",
    val state: String = "MONITORING",
    val crisis_type: String = "",
    val severity_level: Int = 1,
    val confidence_score: Double = 0.0,
    val conflicting_hypothesis: ConflictingHypothesis? = null,
    val location: IncidentLocation = IncidentLocation(),
    val affected_population_estimate: Int = 0,
    val expected_duration_hours: Double = 2.0,
    val peak_impact_time: String? = null,
    val spread_risk: String = "low",
    val severity_forecast: SeverityForecast? = null,
    val resources_allocated: List<String> = emptyList(),
    val stakeholder_notifications_sent: List<String> = emptyList(),
    val signal_ids: List<String> = emptyList(),
    val response_actions: List<Map<String, Any>> = emptyList(),
    val audit_log: List<AuditLogEntry> = emptyList(),
    val trade_off_narrative: String? = null,
    val created_at: String = "",
    val updated_at: String = "",
) {
    /** Severity mapped to a human-readable label. */
    val severityLabel: String
        get() = when (severity_level) {
            1 -> "Minor"
            2 -> "Moderate"
            3 -> "Significant"
            4 -> "Severe"
            5 -> "Catastrophic"
            else -> "Unknown"
        }

    /** True for terminal states that no longer need active monitoring. */
    val isTerminal: Boolean
        get() = state == "RESOLVED" || state == "RETRACTED"
}

data class IncidentLocation(
    val lat: Double = 0.0,
    val lng: Double = 0.0,
    val area_name: String = "",
    val affected_radius_km: Double = 1.0,
)

data class ConflictingHypothesis(
    val type: String = "",
    val confidence: Double = 0.0,
    val evidence_signal_ids: List<String> = emptyList(),
    val evidence_summary: String = "",
)

data class SeverityForecast(
    val t_plus_1h: Int = 1,
    val t_plus_2h: Int = 1,
    val t_plus_6h: Int = 1,
    val uncertainty_range: Int = 1,
)

data class AuditLogEntry(
    val timestamp: String = "",
    val action: String = "",
    val from_state: String? = null,
    val to_state: String? = null,
    val reason: String = "",
    val agent: String = "",
)

File: ciro/schemas/incident.py
# schemas/incident.py
# TEAM CONTRACT — DO NOT CHANGE WITHOUT CONSENSUS
from pydantic import BaseModel, Field
from typing import Optional, Any
from datetime import datetime
import uuid


class IncidentLocation(BaseModel):
    lat: float
    lng: float
    area_name: str
    affected_radius_km: float = 1.0


class ConflictingHypothesis(BaseModel):
    type: str
    confidence: float
    evidence_signal_ids: list[str] = []
    evidence_summary: str = ""


class SeverityForecast(BaseModel):
    t_plus_1h: int  # 1-5
    t_plus_2h: int
    t_plus_6h: int
    uncertainty_range: int = 1


class AuditLogEntry(BaseModel):
    timestamp: str
    action: str
    from_state: Optional[str] = None
    to_state: Optional[str] = None
    reason: str
    agent: str


class Incident(BaseModel):
    incident_id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    state: str = "MONITORING"  # MONITORING | HYPOTHESIS | VERIFICATION_REQUESTED | CONFIRMED | RETRACTED | RESOLVED
    crisis_type: str  # flood | heatwave | accident | infrastructure_failure | power_outage | water_main_burst | road_blockage
    severity_level: int = 1  # 1-5
    confidence_score: float = 0.0
    conflicting_hypothesis: Optional[ConflictingHypothesis] = None
    location: IncidentLocation
    affected_population_estimate: int = 0
    expected_duration_hours: float = 2.0
    peak_impact_time: Optional[str] = None
    spread_risk: str = "low"  # low | medium | high
    severity_forecast: Optional[SeverityForecast] = None
    resources_allocated: list[str] = []
    stakeholder_notifications_sent: list[str] = []
    signal_ids: list[str] = []
    response_actions: list[dict] = []
    audit_log: list[AuditLogEntry] = []
    created_at: str = Field(default_factory=lambda: datetime.utcnow().isoformat())
    updated_at: str = Field(default_factory=lambda: datetime.utcnow().isoformat())

    def to_firestore_dict(self) -> dict[str, Any]:
        data = self.model_dump()
        data["updated_at"] = datetime.utcnow().isoformat()
        return data

    @classmethod
    def from_firestore_dict(cls, data: dict[str, Any]) -> "Incident":
        return cls(**data)
